### PR TITLE
[Ansor][AutoTVM v2.0] Phase 1: XGBoost Cost Model

### PR DESCRIPTION
**Pull request by merrymercy**
For the full upstream plan, see [Ansor RFC](https://discuss.tvm.ai/t/rfc-ansor-an-auto-scheduler-for-tvm-autotvm-v2-0/7005/32).

This PR adds a xgboost-based cost model.
It is similar to the existing xgboost model in autotvm but works on the more general feature representation introduced by #6190 .

RMSE is used as the loss function, but the general feature representation needs slight modification to the loss function.
To support the new loss function, this PR implements a custom xgboost loss function "pack-sum-RMSE".
It is called "pack-sum" because we combine several samples into a "pack" and sum up their predictions.

### CODE REVIEW

**Comment by merrymercy** [comment]
cc @jcf94 @comaniac @junrushao1994 @tqchen

File: python/tvm/auto_scheduler/auto_schedule.py [hunk 1]
```diff
@@ -161,7 +161,7 @@ def __init__(self, task, schedule_cost_model=RandomModel(), params=None, seed=No
             seed or random.randint(1, 1 << 30), verbose, init_search_callbacks)
 
     def generate_sketches(self, print_for_debug=False):
-        """ Generate the sketches, this is mainly used for debug.
+        """ Generate the sketches. This is mainly used for debugging and testing.
```
Comment (comaniac) [inline python/tvm/auto_scheduler/auto_schedule.py:164]: I know the meaning of this description, but it may confuse people. Maybe we can either don't say this is mainly for debugging and testing, or explicitly say this is mainly for debugging and testing because auto scheduler uses them in the C++ side?

File: python/tvm/auto_scheduler/cost_model/xgb_model.py [hunk 2]
```diff
@@ -0,0 +1,590 @@
...
+        """
+        Get an attribute of a xgb.DMatrix
+
+        Parameters
+        ----------
+        key: str
+            The name of the attribute
+        matrix: xgb.DMatrix
+            The matrix
+        default: Optional
```
Comment (comaniac) [inline python/tvm/auto_scheduler/cost_model/xgb_model.py:52]:
```suggestion
        default: Optional[Any]
```

File: python/tvm/auto_scheduler/cost_model/xgb_model.py [hunk 3]
```diff
@@ -0,0 +1,590 @@
...
+        """
+        Set an attribute for a xgb.DMatrix
+
+        Parameters
+        ----------
+        key: str
+            The name of the attribute
+        matrix: xgb.DMatrix
+            The matrix
+        value: Optional
```
Comment (comaniac) [inline python/tvm/auto_scheduler/cost_model/xgb_model.py:67]:
```suggestion
        value: Optional[Any]
```

File: python/tvm/auto_scheduler/cost_model/xgb_model.py [hunk 4]
```diff
@@ -0,0 +1,590 @@
...
+        self.context_dict[key][matrix.handle.value] = value
+
+dmatrix_context = XGBDMatrixContext()
+
+
+class XGBModel(PythonBasedModel):
+    """Train a XGBoost model to predict the normalized throughputs of programs.
+
+    Let the normalized throughput be the score of a program (higher is better). We predict
+    (approximiate) the score of a program = the sum of the scores of all stages in this program.
```
Comment (comaniac) [inline python/tvm/auto_scheduler/cost_model/xgb_model.py:79]:
```suggestion
    the (approximated) score of a program = the sum of the scores of all stages in this program.
```

File: python/tvm/auto_scheduler/cost_model/xgb_model.py [hunk 5]
```diff
@@ -0,0 +1,590 @@
...
+        """Update the cost model according to new measurement results (training data).
+
+        Parameters
+        ----------
+        inputs : List[MeasureInput]
+            The measurement inputs
+        results : List[MeasureResult]
+            The measurement results
+        """
+        if len(inputs) <= 0:
```
Comment (comaniac) [inline python/tvm/auto_scheduler/cost_model/xgb_model.py:130]: We should also check/assert `len(inputs) == len(results)`.

File: python/tvm/auto_scheduler/cost_model/xgb_model.py [hunk 6]
```diff
@@ -0,0 +1,590 @@
...
+
+        super().__init__()
+
+        # cache measurement input/result pairs and extracted features
+        self.inputs = []
+        self.results = []
+        self.inputs_feature_cache = []
+
+    def update(self, inputs, results):
+        """Update the cost model according to new measurement results (training data).
```
Comment (comaniac) [inline python/tvm/auto_scheduler/cost_model/xgb_model.py:121]: 1. It would be better to mention we will train a new model from scratch every time `update` is invoked.
2. This function assumes that there won't be duplicated inputs because we guarantee the same schedule will be measured only once. It would be better to mention it as well.

File: python/tvm/auto_scheduler/cost_model/xgb_model.py [hunk 7]
```diff
@@ -0,0 +1,590 @@
...
+
+    def load_log_file(self, file_name, n_lines=None):
+        """Load measure records from a log file to pre-train the cost model
+
+        Parameters
+        ----------
+        file_name: str
+            The filename
+        n_lines: int
+            Only
```
Comment (comaniac) [inline python/tvm/auto_scheduler/cost_model/xgb_model.py:258]: - s/int/Optional[int]/
- The description seems incomplete.

File: python/tvm/auto_scheduler/cost_model/xgb_model.py [hunk 8]
```diff
@@ -0,0 +1,590 @@
...
+                (np.random.uniform(0, 1, (len(states), )), np.zeros(len(states), )))
+
+        # Predict 0 for invalid states that failed to be lowered.
+        for idx, feature in enumerate(features):
+            if feature.min() == feature.max() == 0:
+                breakdown[idx] = float('-inf')
+
+        return breakdown
+
+    def load_log_file(self, file_name, n_lines=None):
```
Comment (comaniac) [inline python/tvm/auto_scheduler/cost_model/xgb_model.py:250]: From the functionality, it might be better to name it `update_from_file` or something like that.

File: python/tvm/auto_scheduler/cost_model/xgb_model.py [hunk 9]
```diff
@@ -0,0 +1,590 @@
...
+    Parameters
+    ----------
+    preds: np.ndarray
+        The predicitons
+    dtrain: xgb.DMatrix
+        The training set
+
+    Returns
+    -------
+    gradient and hessian
```
Comment (comaniac) [inline python/tvm/auto_scheduler/cost_model/xgb_model.py:402]: Type missing.

File: python/tvm/auto_scheduler/cost_model/xgb_model.py [hunk 10]
```diff
@@ -0,0 +1,590 @@
...
+    pack_ids = dmatrix_context.get("pack_ids", dtrain)
+    weight = dtrain.get_weight()
+
+    sum_pred = np.bincount(pack_ids, weights=preds)
+    x = sum_pred[pack_ids]
+    y = dtrain.get_label()
+    gradient = x - y
+    hessian = np.ones_like(gradient)
+
+    if len(weight) == 0:
```
Comment (comaniac) [inline python/tvm/auto_scheduler/cost_model/xgb_model.py:413]:
```suggestion
    if not weight:
```

File: python/tvm/auto_scheduler/cost_model/xgb_model.py [hunk 11]
```diff
@@ -0,0 +1,590 @@
...
+    Parameters
+    ----------
+    raw_preds: np.ndarray
+        The raw prediction
+    labels: xgb.DMatrix
+        The groud-truth label matrix
+
+    Returns
+    -------
+    The name and value of the metric
```
Comment (comaniac) [inline python/tvm/auto_scheduler/cost_model/xgb_model.py:430]: Type missing.

File: python/tvm/auto_scheduler/cost_model/xgb_model.py [hunk 12]
```diff
@@ -0,0 +1,590 @@
...
+        Parameters
+        ----------
+        raw_preds: np.ndarray
+            The raw prediction
+        labels: xgb.DMatrix
+            The groud-truth label matrix
+
+        Returns
+        -------
+        The name and value of the metric
```
Comment (comaniac) [inline python/tvm/auto_scheduler/cost_model/xgb_model.py:461]: Type missing.

File: python/tvm/auto_scheduler/cost_model/xgb_model.py (hunk 10 above)
Comment (merrymercy) [reply python/tvm/auto_scheduler/cost_model/xgb_model.py:413]: `weight` is an np.ndarray. We can not use `if not weight` for an np.ndarray

**Comment by merrymercy** [comment]
@comaniac  @tqchen comments are addressed.
